### PR TITLE
[WIP] Convert to TensorFlow 2

### PR DESCRIPTION
**Pull request by peastman**
This is in the very early stages.  There are still a *lot* of tests that fail, but on the other hand, there also are a lot that pass, so that's something!

As much as possible, I'm trying to replace calls to functions that have been eliminated with their modern equivalents.  Only when I can't find any modern equivalent, I fall back on the `tf.compat.v1` packages.

I found a couple of places where we were still using `tf.contrib` features.  One was a function that has moved to the tensorflow-probability package, so I added that as a dependency.  The other was the PowerSign optimizer which has been completely eliminated, so I removed support for it.

In TensorFlow 2, Tensor objects are no longer hashable, which means they can't be used as keys in dicts anymore.  As a workaround for that, they allow you to get a hashable reference to a Tensor by calling the `experimental_ref()` method.  As the name suggests, that feature is still experimental and may change in the future.  I'm using it anyway because I don't see any other clean way of handling this.  We'll need to keep an eye on what happens with it in future versions.

I'm not making any effort to convert TensorGraph.  The plan is to remove it in the next release.  That means we need to decide what to do about the remaining models that are based on it, as dicussed in #1732.

### CODE REVIEW

**Comment by peastman** [comment]
I'm having a terrible time getting A3C to work.  A worker thread computes gradients for its local copy of the model, then tries to apply those gradients to the main copy, and nothing happens.  It doesn't change.  I have no idea why.  If instead I use a single thread working with a single copy of the model, everything works.  Something is broken in trying to apply gradients from one model to a different one, but all my attempts at figuring it out have been unsuccessful.

I'm starting to think about just ripping out the threading and switching from A3C to A2C.  Of course, I may run into a similar problem with PPO, but one step at a time.

**Comment by peastman** [comment]
@rbharath any thoughts on this?

**Comment by rbharath** [comment]
I'm unfortunately facing some computer problems at present so can't run the code myself for a bit, but no objections on my end to switching to A2C.

**Comment by peastman** [comment]
I'm trying to figure out what to do about these lines in DAGLayer:

https://github.com/deepchem/deepchem/blob/7492d910100888c07e13fc5286a48b8d67d09638/deepchem/models/layers.py#L2445-L2447

It creates a new variable every time the layer is called. That's an odd thing to do, and it doesn't work correctly with function tracing in TensorFlow 2.  Is there really a reason for it to create a new variable every time instead of just reusing the same variable?

@miaecle

**Comment by peastman** [comment]
I'm trying to get GraphConvModel working.  Debugging `TestGraphModels.test_graph_conv_model()`, I've tracked the problem down to something that looks to me likely clearly incorrect code.  But for some reason that doesn't prevent it from working correctly under TF 1.x, so I'm kind of confused.

When we construct the tf.keras.Model, we give it a list of 16 inputs to expect:

https://github.com/deepchem/deepchem/blob/7492d910100888c07e13fc5286a48b8d67d09638/deepchem/models/graph_models.py#L618-L620

But when we invoke it, we only provide 15 inputs:

https://github.com/deepchem/deepchem/blob/7492d910100888c07e13fc5286a48b8d67d09638/deepchem/models/graph_models.py#L645-L650

There seems to be an inconsistency in the handling of the degree adjacency lists.  Notice that in the above the `range()` starts from 1.  That is, it omits the first list.  But in the list of expected inputs for the model, we included all of them, not omitting the first one.  I don't understand what's going on here.

**Comment by rbharath** [comment]
I think it might have to do with this line

https://github.com/deepchem/deepchem/blob/7492d910100888c07e13fc5286a48b8d67d09638/deepchem/models/graph_models.py#L568

I think the `i+1` means that we do implicitly start from 1, but the code is just very confusingly structured.

**Comment by peastman** [comment]
If I just change it to create one fewer input (`range(10)` instead of `range(0, 10 + 1)`) then it still passes on TF1.  On TF2 it gets a little further, but then fails because of the error described at https://github.com/tensorflow/tensorflow/issues/33425.

**Comment by peastman** [comment]
This is getting close to complete.  I have most features working, with a few exceptions:

- DAGModel and GraphConvModel are both currently broken, as described above.
- I'm not sure what to do about the `in_silico_mutagenesis()` function.  It has this comment about the model you pass to it:

> Currently only SequenceDNN will work, but other models may be added.

SequenceDNN is one of the TensorGraph based models we decided to get rid of.  Should I also remove this function?  Or should it be adapted to work with other models?

- I haven't tested any of the examples yet, but probably some of them will need changes.

**Comment by peastman** [comment]
As far as I can tell, the only thing in `in_silico_mutagenesis()` that's specific to that model is that it expects a field called `num_tasks`.  No other model has a field with that name, but lots of them have fields called `n_tasks`.

In addition, the parent Model class defines an abstract method `get_num_tasks()`.  The default implementation throws `NotImplementedError`, and the only class that implements it is SklearnModel.  How about if I change the default implementation to look for a field called `n_tasks` and return it if it exists, and then have `in_silico_mutagenesis()` call that?

**Comment by rbharath** [comment]
@peastman Sounds good to me!

**Comment by peastman** [comment]
Actually, the real issue isn't about querying the number of tasks.  It's just that the function has very specific expectations about the shapes of the model's inputs and outputs.

**Comment by peastman** [comment]
I think we're now at the point where this can be merged, if you're willing to accept the problems described above.  We can keep working on them in separate PRs.

**Comment by rbharath** [comment]
@peastman Awesome! I'll go ahead and merge this in. We can continue working on the issues in future PRs. I'll open an issue later to coordinate the work that needs to be done.

**Comment by kyzhouhzau** [comment]
This problem still happened in tensorflow-gpu==2.0.0. But, it's ok in tensorflow-gpu==2.1.0.
